Fix feedback support in simple mode (mode 1)

skiselev opened the pull request:
This patch fixes the feedback support in simple mode (mode 1) and addresses https://github.com/daveho/GALasm/issues/12
When configured in simple mode:

- GAL16V8 supports feedback on all OLMC cells except of ones connected to pin 15 and 16
- GAL20V8 supports feedback on all OLMC cells except of ones connected to pin 18 and 19

This patch removes the check that forces complex mode (mode 2) if one of the output pins is used for feedback.
It adds a check for the simple mode output-only pins listed above, so that if they are used for feedback, galasm will use complex mode (mode 2)

skiselev committed:
GALs can do feedback in simple mode - fix it

Signed-off-by: Sergey Kiselev <skiselev@gmail.com>

skiselev committed:
Add verbose output for GAL mode selection

Signed-off-by: Sergey Kiselev <skiselev@gmail.com>

skiselev commented:
The second patch adds verbose output for mode selection by galasm, so that the user can understand better why galasm picked certain mode

pjdennis commented on src/galasm.c:
```diff
@@ -575,18 +575,6 @@ int AssemblePldFile(char *file, struct Config *cfg)
                                 }
                             }
                         }
-
```
Deletion of this block of code looks like it might have unintended consequences for gal_type other than GAL16V8 or GAL20V8, i.e. GAL20RA10 or GAL22V10 - Jedec.GALSYN and Jedec.GALAC0 could be left uninitialized?

pjdennis commented on src/galasm.c:
```diff
@@ -575,18 +575,6 @@ int AssemblePldFile(char *file, struct Config *cfg)
-
-                        /* output and feedback? then mode 2 */
-
-                        if (OLMC[n].PinType == COM_TRI_OUT && OLMC[n].FeedBack)
-                        {
-                            modus = MODE2;      /* mode 2 */
-
-                            Jedec.GALSYN = 1;   /* set SYN,  AC0 bit */
-                            Jedec.GALAC0 = 1;
-
-                            break;
-                        }
```
Deletion of this block of code looks like it might have unintended consequences for gal_type other than GAL16V8 or GAL20V8, i.e. GAL20RA10 or GAL22V10 - Jedec.GALSYN and Jedec.GALAC0 could be left uninitialized?

pjdennis replied:
After a second look, there is a block of code just below which will set 'modus' and initialize Jedec.GALSYN and Jedec.GALAC0, however 'modus' will be set to MODE1 instead of MODE2 and Jedec.GALAC0 will be set to 0 instead of 1.

Jedec.GALAC0 is used only for GAL16V8 and GAL20V8 (in jedec.c) so any difference in value for other GAL types should not cause a problem.

However the potential difference in 'modus' for GAL20RA10 and GAL22V10 looks like it would affect the subsequent loop which is setting PinType and TriCon based on 'modus'. I'm trying to understand if this possible change in behavior for GAL20RA10 and GAL22V10 is desirable.

pjdennis commented:
The first commit (0a3e09ddb9cadf8edb322285d71d5d6d90518eb1) looks good and based on surrounding if statements, it affects only the GAL16V8 and GAL20V8. It makes sense based on the datasheet. I have a GAL16V8 arriving in a few days so I'll test out the change with that chip then.

For the second commit (a967ad526b727de79ba6aa235c69fc515cabca44) I think the new verbose output should be more oriented to the galasm user vs. developer so:
- Consistently reference pin number in the messages vs. sometimes pin number and sometimes OLMC number.
- Not necessary to show the internal mode number (1, 2, 3) in addition to the names (simple, complex, registered)
- Ideally, show all the pins that would force a particular mode (vs. just the first one)

pjdennis committed:
Improving diagnostics

- Show pin numbers and names in verbose diagnostics
- Show all pins causing a mode selection instead of just the first
- Show mode name instead of number - mode number is not meaningful outside the context of the assembler

pjdennis reviewed (approved):
I tested the new simple mode feedback support using a Lattice GAL20V8B and confirmed feedback and input from OLMC pins is working as expected in simple mode. Fallback to complex mode is working when attempting to use the 2 middle OLMC pins for output with feedback or input.

I also modified the diagnostics ([86ea1d3](https://github.com/daveho/GALasm/pull/13/commits/86ea1d382f0e396d0c986719c16ce6cb79e691a2)) as I described in a [prior comment](https://github.com/daveho/GALasm/pull/13#issuecomment-1059903261).

pjdennis commented:
Tested successfully; merging.
